### PR TITLE
Fix ß bug in regexp for mentions and tags

### LINKED ISSUE #31021: Mentioned handles are rendered as plaintext without link if the sharp s ('ß') character appears in the previous word
### Steps to reproduce the problem

1. Write a post with the following content and send it

`Gemäß @Gargron@mastodon.social`

### Expected behaviour

The handle is rendered with a link to the profile: Gemäß @Gargron

### Actual behaviour

The handle is rendered as plaintext: Gemäß @Gargron@mastodon.social

### Detailed description

Apparently when a handle is mentioned after a "ß" (sharp s) character it is always rendered as plaintext.

Please find the following post where it was first observed:
https://nrw.social/@melsdung/112784007422571926

The issue could also be reproduced with the 3rd party app "Tusky" using mastodon.online.

### Mastodon instance

mastodon.online and nrw.social

### Mastodon version

v4.2.10 (nrw.social) and v4.3.0-nightly.2024-07-11 (mastodon.online)

### Technical details

If this is happening on your own Mastodon server, please fill out those:
- Ruby version: (from `ruby --version`, eg. v3.1.2)
- Node.js version: (from `node --version`, eg. v18.16.0)

### PR DESCRIPTION
**Pull request by adamniedzielski**
Fixes #31021

### CODE REVIEW

**Comment by adamniedzielski** [comment]
Screenshot 2024-07-24 at 15 47 12

LOL, ß is the nemesis 😆

File: app/models/account.rb [hunk 1]
```diff
@@ -68,7 +68,7 @@ class Account < ApplicationRecord
   INSTANCE_ACTOR_ID = -99
 
   USERNAME_RE   = /[a-z0-9_]+([a-z0-9_.-]+[a-z0-9_]+)?/i
-  MENTION_RE    = %r{(?<![=/[:word:]])@((#{USERNAME_RE})(?:@[[:word:].-]+[[:word:]]+)?)}i
+  MENTION_RE    = %r{(?<![=/[:word:]])@((#{USERNAME_RE})(?:@[[:word:].-]+[[:word:]]+)?)}
```
Comment (adamniedzielski) [inline app/models/account.rb:71]: The cases insensitive mode (`i`) is not needed here, because it's enabled for `USERNAME_RE`. According to the documentation (https://ruby-doc.org/3.2.2/Regexp.html#class-Regexp-label-Options)

> i, m, and x can also be applied on the subexpression level with the (?on-off) construct, which enables options on, and disables options off for the expression enclosed by the parentheses

And this magically works when including one regexp in another 🎉

File: app/models/tag.rb [hunk 2]
```diff
@@ -37,7 +37,7 @@ class Tag < ApplicationRecord
   HASHTAG_LAST_SEQUENCE = '([[:word:]_]*[[:alpha:]][[:word:]_]*)'
   HASHTAG_NAME_PAT = "#{HASHTAG_FIRST_SEQUENCE}|#{HASHTAG_LAST_SEQUENCE}"
 
-  HASHTAG_RE = %r{(?<![=/)\p{Alnum}])#(#{HASHTAG_NAME_PAT})}i
+  HASHTAG_RE = %r{(?<![=/)\p{Alnum}])#(#{HASHTAG_NAME_PAT})}
```
Comment (adamniedzielski) [inline app/models/tag.rb:40]: I believe that we don't need the case insensitive mode here, because the regexp is constructed from character classes that capture both lower and upper case.

File: spec/models/account_spec.rb [hunk 3]
```diff
@@ -711,6 +711,14 @@
     it 'does not match URL query string' do
       expect(subject.match('https://example.com/?x=@alice')).to be_nil
     end
+
+    it 'matches usernames immediately following the letter ß' do
+      expect(subject.match('Hello toß @alice from me')[1]).to eq 'alice'
+    end
```
Comment (adamniedzielski) [inline spec/models/account_spec.rb:717]: This test is failing on `main`

File: spec/models/account_spec.rb [hunk 4]
```diff
@@ -711,6 +711,14 @@
...
+
+    it 'matches usernames containing uppercase characters' do
+      expect(subject.match('Hello to @aLice@Example.com from me')[1]).to eq 'aLice@Example.com'
+    end
```
Comment (adamniedzielski) [inline spec/models/account_spec.rb:721]: I added this test to make sure that the case insensitive matching is working correctly. This test passes on `main`.

File: spec/models/tag_spec.rb [hunk 5]
```diff
@@ -95,6 +95,10 @@
     it 'does not match purely-numeric hashtags' do
       expect(subject.match('hello #0123456')).to be_nil
     end
+
+    it 'matches hashtags immediately following the letter ß' do
+      expect(subject.match('Hello toß #ruby').to_s).to eq '#ruby'
+    end
```
Comment (adamniedzielski) [inline spec/models/tag_spec.rb:101]: This test is failing on 'main'.

**Comment by adamniedzielski** [comment]
I somehow feel like there may be a bug hidden here, because the issues only happen:
- for `ß` (and not German special characters in general or non-ASCII characters or anything like that). I somehow have the gut feeling that it's related to the fact that `ß` doesn't have the lowercase or uppercase version or something like that
- when the case insensitive mode is enabled for a regex
- when negative lookbehind is used (https://ruby-doc.org/3.2.2/Regexp.html#class-Regexp-label-Anchors)

As if too many negations were causing a subtle bug in Ruby regex logic.

However, I gave up on trying to find a minimum reproduction example which clearly shows the bug, and disabled the case insensitive mode where it's not needed.

**Comment by ThisIsMissEm** [comment]
You might want to reopen this PR with a branch without the special character in the branch name: 
Screenshot 2024-07-24 at 19 11 22

Perhaps `fix/eszett-before-mentions-and-hashtags-breaking`

**Comment by ThisIsMissEm** [comment]
Also, it'd be really good to add tests for a wider array of unicode characters before mentions and hashtags, perhaps you can do a fuzzing test with `#{Faker::String.random(length: 4)} @alice@foo.example` and `#{Faker::String.random(length: 4)} #example`

File: spec/models/account_spec.rb (hunk 4 above)
Comment (oneiros) [reply spec/models/account_spec.rb:721]: I really like this idea. Would you mind adding the same for hashtags?

**Comment by oneiros** [comment]
> I somehow feel like there may be a bug hidden here

I have got a feeling this is related to https://bugs.ruby-lang.org/issues/4044, a bug report that is linked from the `Regexp` documentation.

File: spec/models/account_spec.rb (hunk 4 above)
Comment (adamniedzielski) [reply spec/models/account_spec.rb:721]: Done ✅

Comment (oneiros) [reply spec/models/account_spec.rb:721]: Thanks a lot!

**Review by oneiros — APPROVED** [review]

**Review by ClearlyClaire — APPROVED** [review]
Such a weird bug! Thanks for investigating!
